### ISSUE TITLE
stream_copy_to_stream's $offset argument is unclear

### ISSUE THREAD

**Issue by divinity76** [php/doc-en#2414]
### Description

imo it should have been null-by-default since the beginning, only seeking if the user want it to seek for some reason, but i suspect that would be too much of a BC-break to fix today, however, $offset should be nullable, i should be able to do 
```php
stream_copy_to_stream(... offset:null);
```
instead of
```php
stream_copy_to_stream(... offset: ftell($in));
```

**Comment by nielsdos** [comment]
I transferred this issue to the documentation repo.

The TL;DR is basically in the comment above this one: the behaviour of $offset is unclear in the docs because it is not consistent with the other $offset in stream_get_content. The documentation should clarify that the $offset is relative to *the current* file offset.

RE: tmp files. Right, I forgot about this :sweat_smile: Thanks for reminding me :p

**Comment by smares** [comment]
The offset is not relative to the current offset. If I have a 12 bytes stream and use `stream_get_contents`, the pointer is at the end of the file. `stream_copy_to_stream` without offset or length returns 0 and nothing is copied. If I specify 1 as offset, it returns 11 and copies all but the first byte.

**Comment by damianwadley** [comment]
Yeah, `$offset` does a `SEEK_SET`. But also only if it's >0. Not sure why it wasn't made nullable like `$length`...
https://github.com/php/php-src/blob/php-8.4.3/ext/standard/streamsfuncs.c#L518

**Comment by anonyco** [comment]
Here's a permalink to the lines mentioned by @damianwadley for future viewers to reference:

https://github.com/php/php-src/blob/914f9ad49be800e887b4c5df637ff9af02eb5eeb/ext/standard/streamsfuncs.c#L498-L500

**TL;DR:** `stream_copy_to_stream($from, $to, $length, $offset)` is equivalent to `fseek($from, $offset, SEEK_SET)` followed by  `stream_copy_to_stream($from, $to, $length, $offset)` followed by `stream_copy_to_stream($from, $to, $length, $offset)` IF (and only IF) `$length > 0`. See https://www.php.net/manual/en/function.fseek.php for further reading.

**Gotcha 1:** Beware using a variable `$length` with a previously used stream! If your `$length` variable is 0 but the stream was previously used/seeked to a non-zero position, then `stream_copy_to_stream` will ignore your `$length` and copy from the current position.

**Gotcha 2:** Failure to seek the specified location is a warning, not an error. Keep in mind only opened files on a storage medium can typically be seeked arbitrarily. Pipes like stdout/stdin cannot be seeked.
